### PR TITLE
refactor: proxy callbacks

### LINKED ISSUE #10: Add proxy callback logic
The best way to handle every webhook' payload in a type-safe way is to use our proxy callback logic that we use in other plugins. 

- cleanest and most effective way to handle payload types
- removes the need for individual webhook typeguards

### PR DESCRIPTION
**Pull request by Keyrxng**
Resolves #10
Required by https://github.com/ubiquity-os/plugins-wishlist/issues/45

This PR adds the `proxyCallback` logic that we use in other plugins. I'm adding it because it's the cleanest most effective way to handle type-safe payloads across every webhook event and I need it for handling `pull_request` events.

---

I've noticed a separate issue regarding `Logs` which is that sometimes we get stack traces, sometimes we don't which is not good. 

```ts
 if (!issue.body || !issue.html_url) {
    throw logger.error("Issue body or URL not found", { issueUrl: issue.html_url });
  }
```

This should generate a stack but instead we get:

image

```ts
  if (!issue.body || !issue.html_url) {
    throw logger.error("Issue body or URL not found", { issueUrl: issue.html_url, error: new Error("Issue body or URL not found") });
  }
```

If we manually create an error in the metadata we get a stack trace but you can clearly see that `error` is an empty object which is wrong which is where my suspicion comes from that the logger is wiping out that property under certain conditions.
image

Will open a task to get it addressed

### CODE REVIEW

**Comment by Keyrxng** [comment]
QA: https://github.com/ubq-testing/ask-plugin/issues/7

**Comment by Keyrxng** [comment]
I'm unable to request review for some reason so gonna tag instead

@gentlementlegen @0x4007 @sshivaditya2019 @sshivaditya

**Comment by 0x4007** [comment]
I've never got my own kernel/plugin set up so I don't have any context on this problem and can't add any value in the form of a review.

**Comment by Keyrxng** [comment]
> I've never got my own kernel/plugin set up so I don't have any context on this problem and can't add any value in the form of a review.

It's the solution to this https://github.com/ubiquity-os/ubiquity-os-kernel/issues/10

File: src/handlers/comment-created-callback.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import { bubbleUpErrorComment } from "../helpers/errors";
+
+export async function issueCommentCreatedCallback(
+  context: Context<"issue_comment.created", SupportedEvents["issue_comment.created"]>
+): Promise<CallbackResult> {
+  const {
+    logger,
+    env: { UBIQUITY_OS_APP_NAME },
+  } = context;
+  const question = context.payload.comment.body;
```
Comment (gentlementlegen) [inline src/handlers/comment-created-callback.ts:14]: Should probably be trimmed

File: src/main.ts [hunk 2]
```diff
@@ -1,4 +1,6 @@
+// @ts-expect-error - no types found
```
Comment (gentlementlegen) [inline src/main.ts:1]: Why are these needed

Comment (Keyrxng) [reply src/main.ts:1]: Sometimes my IDE complains about the types then other times it doesn't and I can't stand red file names but I'll remove. Never happen to you on Windows, I assumed that's what it was lmao.

**Comment by Keyrxng** [comment]
QA: https://github.com/ubq-testing/ask-plugin/issues/7#issuecomment-2437802271

File: src/helpers/callback-proxy.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+export function proxyCallbacks(context: Context): ProxyCallbacks {
+  return new Proxy(callbacks, {
+    get(target, prop: SupportedEventsU) {
+      if (!target[prop]) {
+        context.logger.info(`No callbacks found for event ${prop}`);
+        return { status: 204, reason: "skipped" };
+      }
+      return (async () => {
+        try {
+          return await Promise.all(target[prop].map((callback) => handleCallback(callback, context)));
```
Comment (whilefoo) [inline src/helpers/callback-proxy.ts:42]: wouldn't `Promise.allSettled` make more sense because `.all` will fail even if the first one fails...well I guess it's more how you design your callbacks, sometimes 2 callbacks are dependant on each other so you want first to succeed to run the second one, but sometimes you want them independent

File: src/handlers/comment-created-callback.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+export async function issueCommentCreatedCallback(
+  context: Context<"issue_comment.created", SupportedEvents["issue_comment.created"]>
+): Promise<CallbackResult> {
+  const {
+    logger,
+    env: { UBIQUITY_OS_APP_NAME },
+  } = context;
+  const question = context.payload.comment.body.trim();
+  const slugRegex = new RegExp(`@${UBIQUITY_OS_APP_NAME}`, "gi");
```
Comment (whilefoo) [inline src/handlers/comment-created-callback.ts:15]: just a quick thought: shouldn't we only match the begging of the comment for tag because somebody might tag the bot in the middle of the comment accidentally without wanting to ask a question, but on the other hand some people might ask a question like `how do i assign myself @bot`

File: .github/workflows/compute.yml [hunk 5]
```diff
@@ -46,9 +46,9 @@ jobs:
...
-            OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
-            OPENROUTER_API_KEY: ${{ secrets.OPENROUTER_API_KEY }}
-            UBIQUITY_OS_APP_NAME: ${{ secrets.UBIQUITY_OS_APP_NAME }}
\ No newline at end of file
+          SUPABASE_URL: ${{ secrets.SUPABASE_URL }}
+          SUPABASE_KEY: ${{ secrets.SUPABASE_KEY }}
+          VOYAGEAI_API_KEY: ${{ secrets.VOYAGEAI_API_KEY }}
+          OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
+          OPENROUTER_API_KEY: ${{ secrets.OPENROUTER_API_KEY }}
+          UBIQUITY_OS_APP_NAME: ${{ secrets.UBIQUITY_OS_APP_NAME }}
```
Comment (whilefoo) [inline .github/workflows/compute.yml:54]: I'm seeing this is used in a lot of plugins so I think it would make sense that it gets passed to the plugin from the kernel

Comment (Keyrxng) [reply .github/workflows/compute.yml:54]: Or exported from the SDK maybe if it already exports bot id and node_id I think you said?

File: src/handlers/comment-created-callback.ts (hunk 4 above)
Comment (Keyrxng) [reply src/handlers/comment-created-callback.ts:15]: Good point. I've seen intended usage of it with the invocation within the body, so I'm unsure if it's a design choice or if we should enforce typical slash command structure

File: .github/workflows/compute.yml (hunk 5 above)
Comment (whilefoo) [reply .github/workflows/compute.yml:54]: yeah it exports app id and bot id, but the problem is that it will change in prod/beta or your own kernel

File: src/helpers/callback-proxy.ts (hunk 3 above)
Comment (Keyrxng) [reply src/helpers/callback-proxy.ts:42]: Hmm idk you let me know what to do in this scenario mate. Even though it stops others potentially I think it might be better to stick with `.all`.


---

On the matter of callbacks, we had some misunderstanding with how to properly catch the errors with the proxy you might recall. Well this implementation seems to be the way to handle it.

`runPlugin` returns `proxyCallbacks(context)[eventName]` no try-catch. The Proxy object will post the comment inside it's catch using `bubbleupErrorComment`. Which means callbacks need to either just throw their error or use `bubbleupErrorComment` with the posting param disabled. I've found this to work well so far.

File: .github/workflows/compute.yml (hunk 5 above)
Comment (Keyrxng) [reply .github/workflows/compute.yml:54]: valid point that would be more of a nuisance, kernel passing it is far cleaner.

File: src/helpers/callback-proxy.ts (hunk 3 above)
Comment (whilefoo) [reply src/helpers/callback-proxy.ts:42]: We can leave it for now, it's not that important.
I wonder how this Proxy approach will work with the SDK because with the SDK you provide a function that the SDK will call and the SDK will catch the error and post a comment

Comment (Keyrxng) [reply src/helpers/callback-proxy.ts:42]: I liked your Hono worker approach when I first saw it, is this side of the SDK ready to be used to build a plugin yet, I can try it with https://github.com/ubiquity-os/permit-generation/issues/92#event-14901391331?

`proxyCallbacks(ctx)` returns the callback obj
`proxyCallbacks(ctx)[eventName]` returns a single async function which loops all callbacks for that event in the obj. 

So I think we can treat each event as a single `handler` and you'd be in control of `.all` or `.allSettled` if you have the user pass in the callbacks obj to the `proxyCallbacks` function, currently we pull it from file scope

The fn/promise returned by `proxyCallbacks(ctx)[eventName]` posts the error inside it's catch block rn but if it was thrown in there I'm sure the invoking fn would need to catch it which would be the SDK. Needs confirmed tho as that may have been our problem in the first place, throwing it inside the promise and the proxy interfered it actually throwing it. 

I think we should aim to standardize and support `proxyCallbacks` out of the box because it's the best method found so far to handle every webhook payload at the type-level elegantly.


```
npx tsx t.ts
proxyCallbacks {
  'issue_comment.created': [ [AsyncFunction: notificationsRequiringComments] ],
  'issue_comment.edited': [ [AsyncFunction: notificationsRequiringComments] ],
  'issues.unassigned': [ [AsyncFunction: disqualificationNotification] ],
  'pull_request.review_requested': [ [AsyncFunction: reviewNotification] ]
}
```
```
npx tsx t.ts
proxyCallbacks Promise { <pending> }
```

File: src/handlers/comment-created-callback.ts (hunk 4 above)
Comment (Keyrxng) [reply src/handlers/comment-created-callback.ts:15]: @0x4007 rfc

**Comment by Keyrxng** [comment]
QA: https://github.com/ubq-testing/ask-plugin/issues/11#issuecomment-2442253913

Good to merge? @sshivaditya2019 @gentlementlegen @whilefoo

**Comment by 0x4007** [comment]
It's extremely redundant. This can all be expressed in a sentence or two.

**Comment by Keyrxng** [comment]
> It's extremely redundant. This can all be expressed in a sentence or two.

I don't understand what you are referring to

**Review by whilefoo — APPROVED** [review]
nice to see some code refactoring!

**Review by gentlementlegen — APPROVED** [review]
Don't hesitate to move that to the SDK, it makes things much simpler!

**Comment by Keyrxng** [comment]
> Don't hesitate to move that to the SDK, it makes things much simpler!

Yeah we definitely should build our webhook handler logic into the SDK

I have a couple of tasks I'll close out first but I won't forget about it

**Comment by Keyrxng** [comment]
I can't merge if it's me we are waiting on for this fellas
